### PR TITLE
Extract type cast behaviour

### LINKED ISSUE #1: Extract type cast behaviour
By extracting the type casting to its own class it can be easily replaced when needed. Preferring composition over inheritance etc.

 It becomes better testable as well.

I tried to keep BC by proxying to the TypeCaster and instantiating the TypeCaster in the AbstractFilter::getTypeCaster() method when not set in case FilterManager::filter() method is overridden.

---

Originally posted by @netiul at https://github.com/zfcampus/zf-doctrine-querybuilder/pull/57

### PR DESCRIPTION
**Pull request by zluiten**
This is follow up on zfcampus/zf-doctrine-querybuilder#57

|    Q        |   A
|------------ | ------
| Bugfix      | no
| BC Break    | no
| New Feature | yes
| RFC         | no

### Description

By extracting the type casting to its own class it can be easily replaced when needed. Preferring composition over inheritance etc.

It becomes better testable as well.

I kept BC by proxying to the TypeCaster and instantiating the TypeCaster in the AbstractFilter::getTypeCaster() method when not set in case FilterManager::filter() method is overridden.

### CODE REVIEW

**Comment by zluiten** [comment]
Any idea what's needed to get this in?

**Comment by weierophinney** [comment]
@netiul I've rebased against the new "1.9.x" branch; you'll need to pull from your own remote to get your local checkout up-to-date.

However, tests are failing.

I tried a number of different approaches to fix them, but it seems like there's something fundamentally broken in the testing suite or with the existing dependencies, and after a few hours trying to figure it all out, I need to leave that to others to resolve. Hopefully, if they get resolved, your tests will pass again, and we can merge and release.

**Comment by zluiten** [comment]
@weierophinney Thanks for the work you do and have done! About the tests, it looks like they passed on travis?

**Comment by weierophinney** [comment]
You're right! They hadn't been when I was testing before, but they clearly do now! The only issue is CS, which is not your problem to fix in this case (it's a line-length issue with our copyright lines). I'll get that fixed shortly.

File: src/Filter/ODM/AbstractFilter.php
Comment (weierophinney) [inline src/Filter/ODM/AbstractFilter.php]: I don't like this; it makes understanding the positions of arguments paramount, which gets tricky quickly.

In looking at when this is constructed, it's from calling a plugin manager's `get()` method with the optional second argument, `$options`.

The array of options passed from a plugin manager was originally conceived as an _associative array_. That would be one way to fix this here: have the filter manager pass an associative array with named elements, which would allow this to become:

```php
public function __construct(array $params)
{
        $this->setTypeCaster($params['type_caster'] ?? new TypeCaster());
}
```

File: src/Filter/ODM/AbstractFilter.php
Comment (weierophinney) [inline src/Filter/ODM/AbstractFilter.php]: If you do as I suggest with the constructor, this becomes a moot point.

File: src/Filter/ORM/AbstractFilter.php
Comment (weierophinney) [inline src/Filter/ORM/AbstractFilter.php]: I recognize this constructor existed before. What I'd do in this scenario is change the plugin manager to pass the following:

```php
[
    0 => $this,
    'filter_manager' => $this,
    'type_caster' => $this->resolveTypeCaster(),
]
```

and then here, you would do the following:

```php
public function __construct(array $params)
{

    $this->setFilterManager($this->extractFilterManagerFromConstructorParams($params));
    $this->setTypeCaster($params['type_caster'] ?? new TypeCaster());
}

private function extractFilterManagerFromConstructorParams(array $params)
{
    if (isset($params['filter_manager']) && $params['filter_manager'] instanceof ORMFilterManager) {
        return $params['filter_manager'];
    }
    if (isset($params[0]) && $params[0] instanceof ORMFilterManager) {
        return $params[0];
    }
    throw new InvalidArgumentException('Missing or invalid filter manager provided to ' . __CLASS__);
}
```

File: src/Filter/ORM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ORM/TypeCaster.php:35]: Do we need `int` here as well?

File: src/Filter/ORM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ORM/TypeCaster.php:40]: Do we need `bool` here as well?

File: src/Filter/ORM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ORM/TypeCaster.php:44]: Do we need `double` here as well?

File: src/Filter/TypeCastInterface.php
Comment (weierophinney) [inline src/Filter/TypeCastInterface.php]: Please provide docblock annotations for each parameter. We can also bump to a PHP 7 version with this release, allowing you to add explicit typehints where applicable (e.g., `string $field`, `?string $format = null`, `bool $doNotTypecastDatetime = false`).

File: src/Filter/ODM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ODM/TypeCaster.php:32]: Do we need `integer` here?

File: src/Filter/ODM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ODM/TypeCaster.php:35]: Do we need a case for `bool` here?

File: src/Filter/ODM/TypeCaster.php
Comment (weierophinney) [inline src/Filter/ODM/TypeCaster.php:38]: Do we need a case for `double` here?

**Review by weierophinney — CHANGES_REQUESTED** [review]
I love that you've extracted this, and I think it will add a lot of flexibility!

I've provided some notes on improvements you should make before we merge. Please bump the minimum supported PHP version to 7.2 as part of your changes (update `composer.json`, and remove job configuration from `.travis.yml`), and add typehints whereever possible on **new signatures only**.

Finally, I didn't note it, but your tests should import classes used such as `stdClass` and `DateTime`.

**Comment by zluiten** [comment]
@weierophinney Apologies, I somehow completed missed you having reviewed this. Thanks though, I'll soon process your remarks.

File: src/Filter/ODM/AbstractFilter.php
Comment (zluiten) [reply src/Filter/ODM/AbstractFilter.php]: I don't like it as well, but since it was already done like this and I didn't want to distract from the essence of the changes I mimicked the style :-). Will change though!

File: src/Filter/ORM/AbstractFilter.php
Comment (zluiten) [reply src/Filter/ORM/AbstractFilter.php]: :+1: Will do it exactly like that

File: src/Filter/ORM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ORM/TypeCaster.php:35]: Nope, `int` is not doctrine mapping type.

File: src/Filter/ORM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ORM/TypeCaster.php:40]: Nope, `bool` is not doctrine mapping type.

File: src/Filter/ORM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ORM/TypeCaster.php:44]: Nope, `double` is not doctrine mapping type.

File: src/Filter/TypeCastInterface.php
Comment (zluiten) [reply src/Filter/TypeCastInterface.php]: :+1:

File: src/Filter/ODM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ODM/TypeCaster.php:32]: The Doctrine ODM mapping type for integers is `integer`

File: src/Filter/ODM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ODM/TypeCaster.php:35]: The Doctrine ODM mapping type for boolean types is `boolean`

File: src/Filter/ODM/TypeCaster.php
Comment (zluiten) [reply src/Filter/ODM/TypeCaster.php:38]: `double` is not a Doctrine ODM mapping type.

File: test/Query/Provider/DefaultOrmTest.php
Comment (zluiten) [inline test/Query/Provider/DefaultOrmTest.php:214]: These changes fix a faulty test with PHP 7.4.

**Comment by netiul** [comment]
@weierophinney Thanks again for the review and apologies for the late reply. Incorporated your suggestions and meanwhile fixed a bug as noted [here](https://github.com/laminas-api-tools/api-tools-doctrine-querybuilder/pull/5#discussion_r513013983).

The Travis builds with lowest composer dependencies (`--prefer-lowest`) for 7.3 and 7.4 are (still) disabled. I'm not familiar with reason why those are disabled (I tested, builds do fail). I think it's out of scope of this PR to fix that.

**Comment by zluiten** [comment]
@weierophinney Thanks again for the review and apologies for the late reply. Incorporated your suggestions and meanwhile fixed a bug as noted [here](https://github.com/laminas-api-tools/api-tools-doctrine-querybuilder/pull/5#discussion_r513013983).

The Travis builds with lowest composer dependencies (`--prefer-lowest`) for 7.3 and 7.4 are (still) disabled. I'm not familiar with reason why those are disabled (I tested, builds do fail). I think it's out of scope of this PR to fix that.

I've also added support for immutable date doctrine field variants. That was actually the reason I wanted to extract the type casting behavior so I could replace the TypeCaster with one that supported the immutable date types.

**Comment by zluiten** [comment]
Hey @weierophinney, can you have another look?

**Review by weierophinney — CHANGES_REQUESTED** [review]
Hi, @netiul !

I merged another PR earlier today that prepares the package for PHP 8... and solves the testing issues you were observing as well (mainly by updating a few key dependencies). Since I merged that, I rebased for you... and then discovered issues in that previous patch's phpcs updates, which meant I was fixing phpcs issues... anyways, it looks good; you incorporated all the feedback I had, and we now have running tests.

The only last think I need from you: could you write-up an example for the README.md file demonstrating substituting an alternative type caster implementation? (Or is that even necessary?)

**Comment by zluiten** [comment]
@weierophinney Thanks again. I've put something in the README.

**Review by weierophinney — APPROVED** [review]

**Pull request merged**
